cub/test/catch2_test_device_reduce_deterministic.cu

```diff
@@ -0,0 +1,276 @@
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *     * Neither the name of the NVIDIA CORPORATION nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL NVIDIA CORPORATION BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ ******************************************************************************/
+
+#include "insert_nested_NVTX_range_guard.h"
+// above header needs to be included first
+
+#include <cub/detail/rfa.cuh>
+
+#include <thrust/device_vector.h>
+
+#include <algorithm>
+#include <numeric>
+
+#include "c2h/vector.cuh"
+#include "catch2_test_helper.h"
+#include "catch2_test_launch_helper.h"
+#include "cub/block/block_load.cuh"
+#include "cub/device/dispatch/dispatch_reduce_deterministic.cuh"
+#include "cub/thread/thread_load.cuh"
+#include <catch2/catch.hpp>
+
+// %PARAM% TEST_LAUNCH lid 0:1:2
+
+using float_type_list = c2h::type_list<float, double>;
+
+template <typename InputIteratorT, typename OutputIteratorT, typename NumItemsT>
+CUB_RUNTIME_FUNCTION static cudaError_t DeterministicSum(
+  void* d_temp_storage,
+  size_t& temp_storage_bytes,
+  InputIteratorT d_in,
+  OutputIteratorT d_out,
+  NumItemsT num_items,
+  cudaStream_t stream = 0)
+{
+  CUB_DETAIL_NVTX_RANGE_SCOPE_IF(d_temp_storage, "cub::DeviceReduce::DeterministicSum");
+
+  // Signed integer type for global offsets
+  using OffsetT = cub::detail::choose_offset_t<NumItemsT>;
+
+  // The output value type
+  using OutputT = cub::detail::non_void_value_t<OutputIteratorT, cub::detail::value_t<InputIteratorT>>;
+
+  using InitT = OutputT;
+
+  return cub::detail::DeterministicDispatchReduce<InputIteratorT, OutputIteratorT, OffsetT>::Dispatch(
+    d_temp_storage,
+    temp_storage_bytes,
+    d_in,
+    d_out,
+    static_cast<OffsetT>(num_items),
+    InitT{}, // zero-initialize
+    stream);
+}
+
+template <int NOMINAL_BLOCK_THREADS_4B, int NOMINAL_ITEMS_PER_THREAD_4B>
+struct AgentReducePolicy
```

Comment:
Same. Why is the `Agent*Policy` in a test file? Don't we want it to be interface available?

Reply:
This is to test the determinism of the algorithm. By having different block size and threads per block we try to simulate different order for the reduction (summation) and if it is not deterministic it would fail.